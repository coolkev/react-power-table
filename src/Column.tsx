import * as React from 'react';
import { CellProps, Column, StrictColumn } from './ReactPowerTable';
import { debuglog, makePure } from './utils';

/** @internal */
// export const defaultCellComponent = makePure((props: CellProps<any>) => {
//     debuglog('Render cellComponent column: ' + props.column.key + ' value: ' + props.value);
//     return <div>{props.value}</div>;
// });
// defaultCellComponent.displayName = 'defaultCellComponent';

/** @internal */
export function transformColumn<T>(options: Column<T> | string): StrictColumn<T> {

    debuglog('transformColumn', options);

    const col = typeof options === 'string' ? { fieldName: options } as Column<T> : options;

    const core = getColumnCore(options);
    const { cellProps, headerProps } = getCellAndHeaderProps(col);

    const result = {
        ...col,
        ...core,
        cellProps,
        headerCellProps: headerProps,
        headerComponent: col.headerComponent,
        formatter: col.formatter || null,
    } as StrictColumn<T>;

    if (col.cellComponent) {
        result.cellComponent = col.cellComponent;
        result.cellComponentProps = col.cellComponentProps || ((props) => props);

    } else {

        //result.cellComponent = defaultCellComponent;
        result.cellComponentProps = col.cellComponentProps || ((props) => ({ column: props.column, value: props.value }));

    }

    return result;
}

function getCellAndHeaderProps(options: Column<any>) {

    const cssClass = options.cssClass;

    const headerProps = { style: { textAlign: 'left', whiteSpace: 'nowrap', ...(options.headerCellProps && options.headerCellProps.style) }, ...options.headerCellProps };
    let cellProps: ((props: CellProps<any>) => React.HTMLProps<HTMLTableDataCellElement>);

    const cellStaticProps: React.HTMLProps<HTMLTableDataCellElement> = typeof (options.cellProps) === 'function' ? {} : { ...options.cellProps };
    const cellPropsFunc = typeof (options.cellProps) === 'function' ? options.cellProps : () => null;

    //var cssClassFunc: (row: T) => string;

    if (options.width) {
        cellStaticProps.style = { ...cellStaticProps.style, width: options.width };
        headerProps.style = { ...headerProps.style, width: options.width };

    }

    if (options.maxWidth) {
        cellStaticProps.style = { ...cellStaticProps.style, maxWidth: options.maxWidth };
        headerProps.style = { ...headerProps.style, maxWidth: options.maxWidth };

    }

    if (typeof cssClass === 'function') {
        cellProps = (row) => ({ ...cellStaticProps, ...cellPropsFunc(row), className: cssClass(row) });

    } else if (typeof (cssClass) === 'string') {
        cellStaticProps.className = cssClass;
        cellProps = (row) => ({ ...cellStaticProps, ...cellPropsFunc(row) });
    } else {
        cellProps = (row) => ({ ...cellStaticProps, ...cellPropsFunc(row) });
    }
    if (options.headerCssClass) {
        headerProps.className = options.headerCssClass;
    }

    if (options.textAlign) {
        cellStaticProps.style = { ...cellStaticProps.style, textAlign: options.textAlign };
        headerProps.style = { ...headerProps.style, textAlign: options.textAlign };
    }

    debuglog('getCellAndHeaderProps', options);
    return { cellProps, headerProps };

}

export interface ColumnCore<T> {
    key: string | number;
    field: (row: T) => any;
    fieldName: string;
    headerText: string;
}

export function getColumnCore<T>(col: Column<T> | string): ColumnCore<T> {

    if (typeof (col) === 'string') {
        return {
            key: col,
            field: (row) => row[col],
            fieldName: col,
            headerText: col,

        };
    }

    const { field, fieldName, key } = col;

    if (typeof field === 'function') {
        const actualFieldName = fieldName || getExpression(field);
        return {
            key: key || actualFieldName,
            field,
            fieldName: actualFieldName,
            headerText: col.headerText || actualFieldName,
        };
    }
    if (fieldName) {
        return {
            key: key || fieldName,
            field: (row) => row[fieldName],
            fieldName,
            headerText: col.headerText || fieldName,

        };
    }

    if (!key) {
        throw new Error('Must specify value for "key" if field is not used');
    }
    //    field = _row => null;

    return {
        key,
        field: (_row) => null,
        fieldName: null,
        headerText: col.headerText,

    };

}
/** @internal */
export function getExpression(func: (row: any) => any): string {

    const expr = func.toString();

    const myregexp = /(?:return|\w+ => ).*\.(\w+);?/;
    const match = myregexp.exec(expr);
    if (match != null) {
        return match[1];
    }
    return null;
}

// export function columnsChanged(newColumns: (string | Column)[], prevColumns: (string | Column)[]) {

//     if ((!newColumns && prevColumns) || (newColumns && !prevColumns) || newColumns.length != prevColumns.length) {
//         return true;
//     }

//     //const keys = ['key', 'fieldName', 'headerText'];

//     const changed: number[] = [];
//     //let result = false;

//     for (let i = 0; i < prevColumns.length; i++) {
//         const prevCol = prevColumns[i];
//         const newCol = newColumns[i];

//         if (prevCol != newCol) {
//             //console.log('column changed: ' + (prevColumns[i] as any).key)
//             //return true;
//             changed.push(i);
//         }
//     }

//     if (changed.length > 0) {

//         if (changed.length == prevColumns.length) {

//             throw new Error('all columns changed - this means you are probably doing something wrong');
//             //console.warn('all columns changed - this means you are probably doing something wrong');
//         }
//         // else {
//         //     changed.forEach(i => console.log('column changed: ' + (prevColumns[i] as any).key));
//         // }

//         return true;
//     }
//     return false;
// }
