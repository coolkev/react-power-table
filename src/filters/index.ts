import * as Select from 'react-select';
import { BooleanFilter } from './BooleanFilter';
import { DateFilter } from './DateFilter';
import { DecimalFilter } from './DecimalFilter';
import { FilterDefinition, FilterDefinitionOptionsOrFieldName, OperationDefinition } from './FilterDefinition';
import { IntFilter } from './IntFilter';
import { ListFilter } from './ListFilter';
import { RemoteListFilter, RemoteListOptionProvider } from './RemoteListFilter';
import { StringFilter } from './StringFilter';
import { TimeFilter } from './TimeFilter';

export * from './FilterDefinition';

//type Types = 'string' | 'boolean' | 'date' | 'decimal' | 'int';

const types = {
    string: StringFilter,
    boolean: BooleanFilter,
    date: DateFilter,
    decimal: DecimalFilter,
    int: IntFilter,
    list: ListFilter,
    remotelist: RemoteListFilter,
    timespan: TimeFilter,
};
type types = typeof types & { [key: string]: FilterDefinition<any> };

export function getFilterDefinition(type: 'string', options: FilterDefinitionOptionsOrFieldName): StringFilter;
export function getFilterDefinition(type: 'boolean', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): BooleanFilter & { operations: { 'notnull': OperationDefinition<boolean>, 'null': OperationDefinition<boolean> } };
export function getFilterDefinition(type: 'boolean', options: FilterDefinitionOptionsOrFieldName): BooleanFilter;
export function getFilterDefinition(type: 'date', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): DateFilter & { operations: { 'notnull': OperationDefinition<string>, 'null': OperationDefinition<string> } };
export function getFilterDefinition(type: 'date', options: FilterDefinitionOptionsOrFieldName): DateFilter;
export function getFilterDefinition(type: 'time', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): TimeFilter & { operations: { 'notnull': OperationDefinition<string>, 'null': OperationDefinition<string> } };
export function getFilterDefinition(type: 'time', options: FilterDefinitionOptionsOrFieldName): TimeFilter;
export function getFilterDefinition(type: 'decimal', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): DecimalFilter & { operations: { 'notnull': OperationDefinition<number>, 'null': OperationDefinition<number> } };
export function getFilterDefinition(type: 'decimal', options: FilterDefinitionOptionsOrFieldName): DecimalFilter;
export function getFilterDefinition(type: 'int', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): IntFilter & { operations: { 'notnull': OperationDefinition<number>, 'null': OperationDefinition<number> } };
export function getFilterDefinition(type: 'int', options: FilterDefinitionOptionsOrFieldName): IntFilter;
export function getFilterDefinition(type: 'list', options: FilterDefinitionOptionsOrFieldName, items: ReadonlyArray<Select.Option>): ListFilter;
export function getFilterDefinition(type: 'remotelist', options: FilterDefinitionOptionsOrFieldName, optionProvider: RemoteListOptionProvider): RemoteListFilter;

export function getFilterDefinition(type: string, options: FilterDefinitionOptionsOrFieldName, items?: ReadonlyArray<Select.Option> | RemoteListOptionProvider) {

    return new types[type](options, items);
}

export const DataTypes = types;
