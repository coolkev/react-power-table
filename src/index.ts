export * from './ReactPowerTable';
export * from './Paging';
export * from './Sorting';
export * from './GridFilters';
export * from './filters/';
export * from './Column';
export * from './utils';

import { PowerTableProps } from './ReactPowerTable';
export type ReactPowerTable<TRow = {}, TExtraProps = {}> = React.ComponentClass<PowerTableProps<TRow, TExtraProps>>;
