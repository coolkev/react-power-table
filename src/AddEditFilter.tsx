import * as React from 'react';
import * as Button from 'react-bootstrap/lib/Button';
import * as Radio from 'react-bootstrap/lib/Radio';
import * as filters from './filters/FilterDefinition';
import { debuglog } from './utils';

/**
 * @internal
 */
export interface AddEditFilterProps {
    filter: filters.FilterDefinition<any>;
    initialOperation: filters.OperationDefinition<any>;
    initialValue?: any;
    onApplyFilter: (filter: filters.AppliedFilter<any>) => void;
}

/**
 * @internal
 */
export interface AddEditFilterState {

    operationKey: string;
    value: any;

}

/**
 * @internal
 */
export class AddEditFilter extends React.PureComponent<AddEditFilterProps, AddEditFilterState> {

    constructor(props: AddEditFilterProps) {
        super(props);

        const value = props.initialValue || props.filter.defaultValue || '';

        this.state = { value, operationKey: props.initialOperation.key };

        this.selectedOperationChange = this.selectedOperationChange.bind(this);
        this.handleFilterValueChange = this.handleFilterValueChange.bind(this);
        this.applyFilter = this.applyFilter.bind(this);

    }

    private selectedOperationChange(e: React.FormEvent<Radio & HTMLInputElement>) {

        const prevOperationKey = this.state.operationKey;
        const operationKey = e.currentTarget.value;
        //const operation = this.props.filter.operations[operationKey];

        const newState: AddEditFilterState = { operationKey, value: this.state.value };

        this.props.filter.onOperationChange(prevOperationKey, newState);
        // if (prevOperationKey === 'between' && operationKey !== prevOperationKey && typeof this.state.value === 'string') {
        //     newState.value = this.state.value.split(' ')[0];
        // }
        this.setState(newState);
    }

    private handleFilterValueChange(value: string) {
        debuglog('AddEditFilter.filterValueChange', value);
        this.setState({
            value,
        });
    }

    private applyFilter() {
        const operation = this.props.filter.operations[this.state.operationKey];

        this.props.onApplyFilter({ filter: this.props.filter, operation, value: this.state.value });
    }
    render() {
        debuglog('AddEditFilter.render', this.props);

        const filter = this.props.filter;

        const { value, operationKey } = this.state;

        const opComponents = Object.keys(filter.operations).map((opKey) => {
            const op = filter.operations[opKey];
            let children: React.ReactChild;
            if (opKey === this.state.operationKey) {

                const SelectedFilterComponent = op.filterComponent || filter.filterComponent;

                if (SelectedFilterComponent) {
                    children = <div style={{ marginLeft: 20 }}><SelectedFilterComponent filter={filter} operation={op} value={value} onValueChange={this.handleFilterValueChange} onEnterKeyPress={this.applyFilter} /></div>;
                }
            }

            return <FilterOperation key={opKey} operation={op} filter={filter} selected={opKey === operationKey} onChange={this.selectedOperationChange}>{children}</FilterOperation>;
        });

        return (
            <div>
                <div style={{ margin: '10px 0' }}><b>{filter.displayName}</b>
                </div>

                {opComponents.map((opComponent) => opComponent)}

                <div style={{ marginTop: 20 }}>
                    <Button bsStyle="primary" bsSize="sm" onClick={this.applyFilter}>Apply Filter</Button>
                </div>

            </div>);
    }
}

/**
 * @internal
 */
export interface FilterOperationProps extends React.Props<any> {
    operation: filters.OperationDefinition;
    filter: filters.FilterDefinition;

    selected: boolean;
    onChange: React.EventHandler<React.FormEvent<Radio & HTMLInputElement>>;
}

/**
 * @internal
 */
export const FilterOperation = (props: FilterOperationProps) => {

    const { operation, filter, onChange, selected, children } = props;

    return (
        <div key={operation.key}>
            <Radio checked={selected} onChange={onChange} value={operation.key}>
                {filter.radioButtonLabel ? filter.radioButtonLabel({ filter, operation }) : operation.displayName}

            </Radio>
            {children}
        </div>
    );

};
