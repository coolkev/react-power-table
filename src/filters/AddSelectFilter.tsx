import * as React from 'react';
import { AddEditFilter } from './AddEditFilter'
import * as FormControl from 'react-bootstrap/lib/FormControl';
import * as PowerTable from "./definitions/FilterDefinition";
import { BackLink } from "./BackLink";


/**
  * @internal
  */
export interface AddFilterProps {
    cancelAddFilter: () => void;

    availableFilters: PowerTable.AvailableFiltersMap;
    appliedFilters: PowerTable.AppliedFilterType[];
    onApplyFilter: (filter: PowerTable.AppliedFilterType) => void;

}


/**
  * @internal
  */
export interface AddFilterState {
    searchText: string;
    selectedFilterKey?: string;
}



/**
  * @internal
  */
export class AddSelectFilter extends React.PureComponent<AddFilterProps, AddFilterState> {

    constructor(props: AddFilterProps) {
        super(props);
        this.state = { searchText: '' };
        this.onSearchTextChanged = this.onSearchTextChanged.bind(this);
        this.backToPrev = this.backToPrev.bind(this);

    }


    private onSearchTextChanged(e: React.FormEvent<FormControl & HTMLInputElement>) {
        this.setState({
            searchText: e.currentTarget.value
        });

    }

    private newFilterSelected(key: string) {
        this.setState({
            selectedFilterKey: key
        });

    }

    private backToPrev() {
        if (this.state.selectedFilterKey) {
            this.setState({ selectedFilterKey: null });
        } else {
            this.props.cancelAddFilter();
        }
    }
    render() {

        const props = this.props;

        const { searchText, selectedFilterKey } = this.state;

        if (selectedFilterKey) {

            const filter = props.availableFilters[selectedFilterKey];

            const initialOperation = filter.operations.all[0];

            return <div>
                <BackLink onClick={this.backToPrev} />

                <div>

                    <AddEditFilter filter={filter} initialOperation={initialOperation} onApplyFilter={this.props.onApplyFilter} />

                </div>
            </div>;

        }

        const unusedFilters = props.availableFilters.all.filter(m => props.appliedFilters.every(c => c.filter.fieldName != m.fieldName));

        const regex = new RegExp(searchText, 'i');
        const availableFilters = searchText ? unusedFilters.filter(m => m.displayName.match(regex)) : unusedFilters;


        return <div className="flex-column">
            <BackLink onClick={this.backToPrev} />


            <div><FormControl placeholder="Filter by" value={searchText} onChange={this.onSearchTextChanged} autoFocus /></div>

            <div className="small flex-column">
                <div style={{ margin: '10px 0' }}><b>Available Filters</b></div>
                <div className="available-filters">
                    <div className="list-group">
                        {availableFilters.map(m => <a href="#" onClick={e => {
                            e.preventDefault();
                            this.newFilterSelected(m.fieldName);
                        }} className="list-group-item" key={m.fieldName}>{m.displayName}</a>)}
                    </div>
                </div>

            </div>

        </div>;
    }
}




