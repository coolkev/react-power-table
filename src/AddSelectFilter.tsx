import * as React from 'react';
import FormControl from 'react-bootstrap/lib/FormControl';
import { AddEditFilter } from './AddEditFilter';
import { BackLink } from './components/BackLink';
import { FilterDefinition, AppliedFilter, OperationDefinition } from './filters/FilterDefinition';
import { objectMapToArray } from './utils';

/**
 * @internal
 */
export interface AddFilterProps {
    cancelAddFilter: () => void;

    availableFilters: FilterDefinition[] | { [key: string]: FilterDefinition };

    //availableFilters: { [key: string]: PowerTable.FilterDefinition };
    appliedFilters: AppliedFilter[];
    onApplyFilter: (filter: AppliedFilter) => void;

    onlyShowUnused?: boolean;

}

/**
 * @internal
 */
export interface AddFilterState {
    searchText: string;
    selectedFilterKey?: string;
}

/**
 * @internal
 */
export class AddSelectFilter extends React.PureComponent<AddFilterProps, AddFilterState> {

    constructor(props: AddFilterProps) {
        super(props);
        this.state = { searchText: '' };
        this.onSearchTextChanged = this.onSearchTextChanged.bind(this);
        this.backToPrev = this.backToPrev.bind(this);
        this.newFilterSelected = this.newFilterSelected.bind(this);

    }

    private onSearchTextChanged(e: React.FormEvent<FormControl & HTMLInputElement>) {
        this.setState({
            searchText: e.currentTarget.value,
        });

    }

    private newFilterSelected(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        this.setState({
            selectedFilterKey: e.currentTarget.dataset.fieldname,
        });
    }

    private backToPrev(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        if (this.state.selectedFilterKey) {
            this.setState({ selectedFilterKey: null });
        } else {
            this.props.cancelAddFilter();
        }
    }
    render() {

        const { availableFilters, appliedFilters, onlyShowUnused } = this.props;

        const { searchText, selectedFilterKey } = this.state;

        if (selectedFilterKey) {

            const filter = Array.isArray(availableFilters) ? availableFilters.find((f) => f.fieldName === selectedFilterKey) : availableFilters[selectedFilterKey];

            const initialOperation = objectMapToArray(filter.operations)[0] as OperationDefinition;

            return (
                <div>
                    <BackLink onClick={this.backToPrev} />

                    <div>

                        <AddEditFilter filter={filter} initialOperation={initialOperation} onApplyFilter={this.props.onApplyFilter} />

                    </div>
                </div>
            );

        }

        const filters = objectMapToArray(availableFilters);

        const visibleFilters = onlyShowUnused ? filters.filter((m) => appliedFilters.every((c) => c.filter.fieldName !== m.fieldName)) : filters;

        const regex = new RegExp(searchText, 'i');
        const showFilters = searchText ? visibleFilters.filter((m) => m.displayName.match(regex)) : visibleFilters;

        // const onSelectFilter = (e: React.) => {
        //     e.preventDefault();
        //     this.newFilterSelected(m.fieldName);
        // };

        return (
            <div className="flex-column">
                <BackLink onClick={this.backToPrev} />

                <div><FormControl placeholder="Filter by" value={searchText} onChange={this.onSearchTextChanged} autoFocus /></div>

                <div className="small flex-column">
                    <div style={{ margin: '10px 0' }}><b>Available Filters</b></div>
                    <div className="available-filters">
                        <div className="list-group">
                            {showFilters.map((m) => <a href="#" onClick={this.newFilterSelected} data-fieldname={m.fieldName} className="list-group-item" key={m.fieldName}>{m.displayName}</a>)}
                        </div>
                    </div>

                </div>

            </div>
        );
    }
}
