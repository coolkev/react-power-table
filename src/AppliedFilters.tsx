import * as React from 'react';
import { ActionLink } from './components/ActionLink';
import * as filters from './filters';

export interface AppliedFiltersProps {
    appliedFilters: ReadonlyArray<filters.AppliedFilter>;
    removeFilter: (filter: filters.AppliedFilter) => void;
    editFilter: (filter: filters.AppliedFilter) => void;

}

export const AppliedFilters = (props: AppliedFiltersProps) => {
    return (
        <div className="applied-filters small">
            {props.appliedFilters.map((appliedFilter) => <AppliedFilter appliedFilter={appliedFilter} key={appliedFilter.key} onEditFilter={props.editFilter} onRemoveFilter={props.removeFilter} />)}
        </div>
    );
};

export interface AppliedFilterProps {
    appliedFilter: filters.AppliedFilter;
    onRemoveFilter: (filter: filters.AppliedFilter) => void;
    onEditFilter: (filter: filters.AppliedFilter) => void;

}

const defaultAppliedFilterLabelComponent = (props: filters.AppliedFilter<any>) => {
    const appliedLabel = props.operation.appliedLabel || props.filter.appliedLabel || filters.FilterDefinition.defaultAppliedFilterLabel;
    return <span>{appliedLabel(props)}</span>;
};

export const AppliedFilter = ({ appliedFilter, onRemoveFilter, onEditFilter }: AppliedFilterProps) => {

    const handleRemoveFilter = React.useCallback((e: React.SyntheticEvent<any>) => {
        e.preventDefault();
        onRemoveFilter(appliedFilter);
    }, [appliedFilter, onRemoveFilter]);

    const handleEditFilter = React.useCallback(() => {

        onEditFilter(appliedFilter);

    }, [appliedFilter, onEditFilter]);

    const AppliedLabelComponent = appliedFilter.operation.appliedLabelComponent || appliedFilter.filter.appliedLabelComponent || defaultAppliedFilterLabelComponent;

    return (
        <div className="applied-filter well well-sm" style={{ marginBottom: 10 }} key={appliedFilter.filter.fieldName}>
            <button type="button" className="close" aria-label="Remove" onClick={handleRemoveFilter}><span aria-hidden="true">Ã—</span></button>
            <ActionLink onClick={handleEditFilter}><AppliedLabelComponent {...appliedFilter} /></ActionLink>
        </div>
    );

};

