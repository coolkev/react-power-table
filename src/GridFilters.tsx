import * as React from 'react';
import { AddEditFilter } from './AddEditFilter';
import { AddSelectFilter } from './AddSelectFilter';
import { AppliedFilters } from './AppliedFilters';
import { BackLink } from './components/BackLink';
import { AppliedFilter, FilterDefinition } from './filters/FilterDefinition';

export interface GridFiltersProps {

    availableFilters: FilterDefinition[] | { [key: string]: FilterDefinition };
    appliedFilters: AppliedFilter[];
    onFiltersChange: (newFilters: AppliedFilter[]) => void;

    onlyShowUnused?: boolean;

}

interface AppliedFilterDTO {
    columnKey: string;
    operationKey: string;
    value: string;
}

interface GridFiltersState {

    addingFilter?: boolean;
    editingFilter?: AppliedFilter;

}

class GridFiltersInternal extends React.Component<GridFiltersProps, GridFiltersState> {

    constructor(props: GridFiltersProps) {
        super(props);
        this.state = {};
        this.hideAddFilter = this.hideAddFilter.bind(this);
        this.applyNewfilter = this.applyNewfilter.bind(this);
        this.removeFilter = this.removeFilter.bind(this);
        this.editFilter = this.editFilter.bind(this);
        this.cancelEditFilter = this.cancelEditFilter.bind(this);
        this.applyEditFilter = this.applyEditFilter.bind(this);
        this.showAddFilter = this.showAddFilter.bind(this);
        this.handleRemoveEditFilter = this.handleRemoveEditFilter.bind(this);
    }

    public render() {

        const { editingFilter } = this.state;
        const { appliedFilters, availableFilters, onlyShowUnused } = this.props;

        if (editingFilter) {
            const { filter, operation, value, key } = editingFilter;

            return (
                <div>
                    <BackLink onClick={this.cancelEditFilter} />

                    <div>

                        <AddEditFilter filter={filter} initialOperation={operation} initialValue={value} onApplyFilter={this.applyEditFilter} onRemoveFilter={this.handleRemoveEditFilter} filterKey={key} />

                    </div>

                </div>
            );
        }

        const children = this.state.addingFilter
            ? <AddSelectFilter cancelAddFilter={this.hideAddFilter} appliedFilters={appliedFilters} availableFilters={availableFilters} onApplyFilter={this.applyNewfilter} onlyShowUnused={onlyShowUnused} />
            : <div><a href="#" onClick={this.showAddFilter}><span className="glyphicon glyphicon-plus" style={{ marginRight: 5 }} />Add Filter</a></div>;

        return (
            <div className="flex-column">

                <AppliedFilters appliedFilters={appliedFilters} removeFilter={this.removeFilter} editFilter={this.editFilter} />

                {children}
            </div>
        );

    }
    private showAddFilter(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        this.setState({
            addingFilter: true,
        });
    }
    private hideAddFilter() {
        this.setState({
            addingFilter: false,
        });
    }

    private applyNewfilter(filter: AppliedFilter) {

        const count = filter.key ? 1 : this.props.appliedFilters.filter(m => m.filter.fieldName === filter.filter.fieldName).length + 1;
        const key = filter.key || filter.filter.fieldName + (count === 1 ? '' : count);
        const newFilters = [...this.props.appliedFilters, { ...filter, key }];

        this.props.onFiltersChange(newFilters);
        this.setState({
            addingFilter: false,
        });
        //}
    }

    private removeFilter(filter: AppliedFilter) {
        const newFilters = this.props.appliedFilters.filter((m) => m !== filter);

        this.props.onFiltersChange(newFilters);
    }

    private handleRemoveEditFilter() {
        const toRemove = this.state.editingFilter;
        const newFilters = this.props.appliedFilters.filter((m) => m !== toRemove);

        this.props.onFiltersChange(newFilters);

        this.setState({
            editingFilter: null
        });
    }
    private editFilter(filter: AppliedFilter) {

        this.setState({
            editingFilter: filter,
            addingFilter: false,
        });
    }

    private cancelEditFilter(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        this.setState({
            editingFilter: null,
        });
    }

    private applyEditFilter(filter: AppliedFilter) {

        const newFilters = this.props.appliedFilters.map((m) => m.key !== filter.key ? m : filter);

        this.props.onFiltersChange(newFilters);
        this.setState({
            editingFilter: null,
        });

    }

}

export const GridFilters: React.ComponentClass<GridFiltersProps> = GridFiltersInternal;
