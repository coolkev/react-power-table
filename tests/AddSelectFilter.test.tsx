import * as React from 'react';
import { GridFilters, DataTypes, AppliedFilter } from '../src/';
import { defaultColumns, sampledata, President } from './shared';
import { mount } from 'enzyme';
import { AddSelectFilter } from '../src/AddSelectFilter';

//const columns = defaultColumns;
const rows = sampledata.slice(0, 25);

const availableFilters = [
    new DataTypes.int('number'),
    new DataTypes.string('president'),
    new DataTypes.int('birth_year'),
    new DataTypes.int({ fieldName: 'death_year', canBeNull: true }),
    new DataTypes.date({ fieldName: 'took_office' }),
    new DataTypes.date({ fieldName: 'left_office', canBeNull: true }),
    new DataTypes.boolean({ fieldName: 'assasinated', displayName: 'was assasinated' }),

];

const noop = () => { };

describe('AddSelectFilter tests',
    () => {

        test('add filter', () => {

            const c = <AddSelectFilter availableFilters={availableFilters} appliedFilters={[]} cancelAddFilter={noop} onApplyFilter={noop} />;

            const component = mount(c);

            expect(component.render()).toMatchSnapshot();

        });

        test('select filter', () => {

            const c = <AddSelectFilter availableFilters={availableFilters} appliedFilters={[]} cancelAddFilter={noop} onApplyFilter={noop} />;

            const component = mount(c);

            expect(component.render()).toMatchSnapshot();

            component.setState({ selectedFilterKey: 'birth_year' });

            expect(component.render()).toMatchSnapshot();

        });

    });
