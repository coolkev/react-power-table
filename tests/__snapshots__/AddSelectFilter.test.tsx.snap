// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSelectFilter tests add filter 1`] = `
<div
  class="flex-column"
>
  <div
    style="margin-bottom: 10px;"
  >
    <a
      href="#"
    >
      <span
        aria-hidden="true"
        class="glyphicon glyphicon-chevron-left"
      />
      Back
    </a>
  </div>
  <div>
    <input
      class="form-control"
      placeholder="Filter by"
      value=""
    />
  </div>
  <div
    class="small flex-column"
  >
    <div
      style="margin: 10px 0px;"
    >
      <b>
        Available Filters
      </b>
    </div>
    <div
      class="available-filters"
    >
      <div
        class="list-group"
      >
        <a
          class="list-group-item"
          data-fieldname="number"
          href="#"
        >
          number
        </a>
        <a
          class="list-group-item"
          data-fieldname="president"
          href="#"
        >
          president
        </a>
        <a
          class="list-group-item"
          data-fieldname="birth_year"
          href="#"
        >
          birth_year
        </a>
        <a
          class="list-group-item"
          data-fieldname="death_year"
          href="#"
        >
          death_year
        </a>
        <a
          class="list-group-item"
          data-fieldname="took_office"
          href="#"
        >
          took_office
        </a>
        <a
          class="list-group-item"
          data-fieldname="left_office"
          href="#"
        >
          left_office
        </a>
        <a
          class="list-group-item"
          data-fieldname="assasinated"
          href="#"
        >
          was assasinated
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`AddSelectFilter tests select filter 1`] = `
<div
  class="flex-column"
>
  <div
    style="margin-bottom: 10px;"
  >
    <a
      href="#"
    >
      <span
        aria-hidden="true"
        class="glyphicon glyphicon-chevron-left"
      />
      Back
    </a>
  </div>
  <div>
    <input
      class="form-control"
      placeholder="Filter by"
      value=""
    />
  </div>
  <div
    class="small flex-column"
  >
    <div
      style="margin: 10px 0px;"
    >
      <b>
        Available Filters
      </b>
    </div>
    <div
      class="available-filters"
    >
      <div
        class="list-group"
      >
        <a
          class="list-group-item"
          data-fieldname="number"
          href="#"
        >
          number
        </a>
        <a
          class="list-group-item"
          data-fieldname="president"
          href="#"
        >
          president
        </a>
        <a
          class="list-group-item"
          data-fieldname="birth_year"
          href="#"
        >
          birth_year
        </a>
        <a
          class="list-group-item"
          data-fieldname="death_year"
          href="#"
        >
          death_year
        </a>
        <a
          class="list-group-item"
          data-fieldname="took_office"
          href="#"
        >
          took_office
        </a>
        <a
          class="list-group-item"
          data-fieldname="left_office"
          href="#"
        >
          left_office
        </a>
        <a
          class="list-group-item"
          data-fieldname="assasinated"
          href="#"
        >
          was assasinated
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`AddSelectFilter tests select filter 2`] = `
<div>
  <div
    style="margin-bottom: 10px;"
  >
    <a
      href="#"
    >
      <span
        aria-hidden="true"
        class="glyphicon glyphicon-chevron-left"
      />
      Back
    </a>
  </div>
  <div>
    <div>
      <div
        style="margin: 10px 0px;"
      >
        <b>
          birth_year
        </b>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            title=""
          >
            <input
              checked=""
              type="radio"
              value="eq"
            />
            is equal to
          </label>
        </div>
        <div
          style="margin-left: 20px;"
        >
          <span
            class="has-error"
          >
            <input
              class="form-control input-sm"
              type="number"
              value=""
            />
          </span>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            title=""
          >
            <input
              type="radio"
              value="ne"
            />
            is not equal to
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            title=""
          >
            <input
              type="radio"
              value="lt"
            />
            is less than
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            title=""
          >
            <input
              type="radio"
              value="gt"
            />
            is greater than
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            title=""
          >
            <input
              type="radio"
              value="between"
            />
            is between
          </label>
        </div>
      </div>
      <div
        style="margin-top: 20px;"
      >
        <button
          class="btn btn-sm btn-primary"
          type="button"
        >
          Apply Filter
        </button>
      </div>
    </div>
  </div>
</div>
`;
